mcwb/volume.py

```diff
@@ -0,0 +1,131 @@
+    ):
+        if (size is None) == (end is None):
+            raise InvalidArgument(
+                "Volume constructor takes only one of size or end")
+        position = Vec3(*position)
+
+        if end is None:
+            size = Vec3(*size)
+            offset = Vec3(0, 0, 0)
+            if anchor.value in Anchor3.TOP.value:
+                offset += Vec3(0, 1 - size.y, 0)
+            if anchor.value in Anchor3.SOUTH.value:
+                offset += Vec3(0, 0, 1 - size.z)
+            if anchor.value in Anchor3.EAST.value:
+                offset += Vec3(1 - size.x, 0, 0)
+            if anchor.value in Anchor3.MIDDLE_FACE.value:
+                offset -= Vec3(int(size.x / 2), 0, int(size.z / 2))
+            elif anchor.value == Anchor3.MIDDLE.value:
+                offset = Vec3(-1, -1, -1) * (size / 2).with_ints()
+
+            self.start = position + offset
+            self.end = self.start + (size - 1)
+        else:
+            end = Vec3(*end)
+            # normalize start(position) and end so all start coords are minima
+            self.start = Vec3(
+                min(position.x, end.x), min(
+                    position.y, end.y), min(position.z, end.z)
+            )
+            self.end = Vec3(
+                max(position.x, end.x), max(
+                    position.y, end.y), max(position.z, end.z)
+            )
+            position = self.start
+            anchor = Anchor3.BOTTOM_NW
+            size = (self.end - position) + 1
+
+        self.size = size
+        self.position = position
+        self.anchor = anchor
+
+    def inside(
+        self, position: Vec3, xtol: int = 0, ytol: int = 0, ztol: int = 0
+    ) -> bool:
+        """ determine if position is within the Volume """
+        return (
+            self.start.x - xtol <= position.x <= self.end.x + xtol
+            and self.start.y - ytol <= position.y <= self.end.y + ytol
+            and self.start.z - ztol <= position.z <= self.end.z + ztol
+        )
+
+    def move(self, distance: Vec3) -> None:
+        """ move the volume's location in space by distance """
+        self.start += distance
+        self.position += distance
+
+    def move_to(self, position: Vec3) -> None:
+        """ move the volume's location in space to position """
+        self.start += position - self.position
+        self.position = position
+
+    def fill(self, client: Client, block: Item = Item.AIR):
+        """ Fill the Volume with a single block type, supports large volumes """
+        if self.size.volume < 32768:
+            client.fill(self.start, self.end, block.value)
+        else:
+            profile = [[block.value] * int(self.size.x)] * int(self.size.y)
+            mktunnel(
+                client,
+                profile,
+                self.start,
+                direction=Direction.UP,
+                anchor=Anchor.BOTTOM_LEFT,
+                length=int(self.size.dz),
+            )
+
+    def walls(
```

Comment:
The function has a lot of args, some of which are single characters.
Please consider better variable names and making some of them keyword-only.

Reply:
I have made the argument changes suggested. 

I'm not sure about having fill() and walls() anyway since I originally conceived Volume as simply describing a space for use in builder functions. It does seem a convenient place to put what are likely to be common builder functions though?